### PR TITLE
Add failover group operations

### PR DESCRIPTION
**Pull request by btasdoven**
Adding Failover Groups operations to Microsoft.Sql provider with their tests, and reflecting the [API changes](https://github.com/Azure/azure-rest-api-specs/pull/1181) in our swagger file.

This is a breaking change in terms of Failover Groups APIs; however, this should not be an issue since the feature has not gone public.

## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as a link to the swagger spec, used to generate the code.
- [ ] The `project.json` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

File: src/SDKs/SqlManagement/Management.Sql/Generated/DatabasesOperations.cs [hunk 1]
```diff
@@ -4003,7 +4003,7 @@ internal DatabasesOperations(SqlManagementClient client)
             HttpStatusCode _statusCode = _httpResponse.StatusCode;
             cancellationToken.ThrowIfCancellationRequested();
             string _responseContent = null;
-            if ((int)_statusCode != 201 && (int)_statusCode != 202)
+            if ((int)_statusCode != 200 && (int)_statusCode != 202)
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Management.Sql/Generated/DatabasesOperations.cs]: Note to reviewers: Swagger change was https://github.com/Azure/azure-rest-api-specs/pull/1166

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                Server partnerServer = CreateServer(sqlClient, resourceGroup, testPrefix, DefaultSecondaryLocation);
+
+                // Create a failover group
+                //
+                string failoverGroupName = SqlManagementTestUtilities.GenerateName(testPrefix);
+                var fgInput = new FailoverGroup()
+                {
+                    ReadOnlyEndpoint = new FailoverGroupReadOnlyEndpoint()
+                    {
+                        FailoverPolicy = "Disabled",
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: You should be able to write `ReadOnlyEndpointFailoverPolicy.Disabled`. (How do I know? see https://github.com/Azure/azure-rest-api-specs/blob/master/arm-sql/2015-05-01-preview/swagger/failoverGroups.json#L490 )

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                    },
+                    ReadWriteEndpoint = new FailoverGroupReadWriteEndpoint()
+                    {
+                        FailoverPolicy = "Automatic",
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Use `ReadWriteEndpointFailoverPolicy.Automatic`

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+                // Delete failover group and verify
+                //
+                sqlClient.FailoverGroups.Delete(resourceGroup.Name, primaryServer.Name, failoverGroupName);
+                Assert.Throws<Microsoft.Rest.Azure.CloudException>(() => sqlClient.FailoverGroups.Get(resourceGroup.Name, primaryServer.Name, failoverGroupName));
+            });
+        }
+
+        [Fact]
+        public void TestUpdateFailoverGroup()
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Seems a bit repetitive, could this be part of the above test case?

Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Or could you at least factor out the common setup code?

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                    },
+                };
+                failoverGroup = sqlClient.FailoverGroups.CreateOrUpdate(resourceGroup.Name, primaryServer.Name, failoverGroupName, fgUpdateInput);
+                SqlManagementTestUtilities.ValidateFailoverGroup(fgUpdateInput, failoverGroup, failoverGroupName);
+
+                var sourceDatabase = sqlClient.Databases.Get(resourceGroup.Name, primaryServer.Name, databaseName);
+                var targetDatabase = sqlClient.Databases.Get(resourceGroup.Name, partnerServer.Name, databaseName);
+                Assert.NotNull(sourceDatabase.FailoverGroupId);
+                Assert.NotNull(targetDatabase.FailoverGroupId);
+
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Could you have a test where you call Failover? Seems like an important scenario for FailoverGroups ;)

Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: It doesn't have to be a separate test, could be part of this or the other one. As long as it's covered somehow.

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Sql.Tests
+{
+    public class FailoverGroupCrudScenarioTests
+    {
+        public static string DefaultPrimaryLocation
```
Comment (btasdoven) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Yup, they are for Stage. Moved them to SqlManagementTestUtilities.

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+        public static string DefaultSecondaryLocation
+        {
+            get
+            {
+                return "SouthEast Asia";
+            }
+        }
+
+        private Server CreateServer(SqlManagementClient sqlClient, ResourceGroup resourceGroup, string serverPrefix, string location)
```
Comment (btasdoven) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Done.

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs (hunk 5 above)
Comment (btasdoven) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: That makes sense :) Added failover operation as part of the test case.

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs (hunk 4 above)
Comment (btasdoven) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Combined the test cases into one.

File: src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+                // Failover failover group
+                //
+                failoverGroup = sqlClient.FailoverGroups.Failover(resourceGroup.Name, targetServer.Name, failoverGroupName);
+
+                // Get failover group on the new secondary server and verify its replication role
+                //
+                var failoverGroupOnSecondary = sqlClient.FailoverGroups.Get(resourceGroup.Name, sourceServer.Name, failoverGroupName);
+                Assert.Equal(FailoverGroupReplicationRole.Secondary, failoverGroupOnSecondary.ReplicationRole);
+                Assert.Equal(FailoverGroupReplicationRole.Primary, failoverGroupOnSecondary.PartnerServers.First().ReplicationRole);
```
Comment (jaredmoo) [inline src/SDKs/SqlManagement/Sql.Tests/FailoverGroupCrudScenarioTests.cs]: Beautiful. 💯

**Review by jaredmoo — APPROVED** [review]

**Review by shahabhijeet — CHANGES_REQUESTED** [review]
I don't see any change in version.
Will need changes to csproj and assemblyinfo files related to versions.

**Comment by shahabhijeet** [comment]
@azuresdkci add to whitelist

**Review by shahabhijeet — APPROVED** [review]

**Pull request merged**